kv_connectors/llmd_fs_backend/src/llmd_fs_backend/worker.py

```diff
@@ -0,0 +1,296 @@
+        list_num_blocks_idx = []
+        list_kv_before_num_blocks = []
+        list_layers_before_num_blocks = []
+
+        for layer_name, gpu_tensor in gpu_caches.items():
+            gpu_shape = gpu_tensor.shape
+            attn_backend = self.attn_backends[layer_name]
+
+            # Generate a reference KV-cache shape using known parameters.
+            # We compare gpu_shape with this synthetic shape to infer the layout.
+            test_shape = attn_backend.get_kv_cache_shape(
+                num_blocks=1234, block_size=16, num_kv_heads=8, head_size=256
+            )
+
+            # Case 1: Cross-layer tensor - an extra layer dimension exists on each tensor.
+            # In this case, num_blocks is the leading dimension.
+            if len(gpu_shape) != len(test_shape):
+                assert len(gpu_shape) == len(test_shape) + 1
+                num_blocks_idx = 0
+                kv_before_num_blocks = False
+                layers_before_num_blocks = False
+
+            # Case 2: Standard layout - each element represents a single layer with
+            # tensor shaped as (num_blocks, ...). The first dimension matches num_blocks.
+            elif test_shape[0] == 1234:
+                num_blocks_idx = 0
+                kv_before_num_blocks = False
+                layers_before_num_blocks = True
+
+            # Case 3: (2, num_blocks, ...) - standard layout but with KV first:
+            # (2, num_blocks, heads, block_size, head_size).
+            else:
+                assert test_shape[0] == 2
+                assert test_shape[1] == 1234
+                assert gpu_shape[0] == 2
+
+                num_blocks_idx = 1
+                kv_before_num_blocks = True
+                layers_before_num_blocks = True
+
+            # Store layout metadata for this layer.
+            list_num_blocks_idx.append(num_blocks_idx)
+            list_kv_before_num_blocks.append(kv_before_num_blocks)
+            list_layers_before_num_blocks.append(layers_before_num_blocks)
+
+        return (
+            list_num_blocks_idx,
+            list_kv_before_num_blocks,
+            list_layers_before_num_blocks,
+        )
+
+# ----------------------------------------------------------------------
+# GPU → Storage (PUT)
+# ----------------------------------------------------------------------
+class GPUStorageOffloadingHandler(StorageOffloadingHandler):
+    """Handler for writing KV blocks from GPU tensors into shared storage."""
+    def __init__(
+        self,
+        model_name: str,
+        tp_size: int,
+        tp_rank: int,
+        kv_caches: Dict[str, torch.Tensor],
+        gpu_blocks_per_file: int,
+        attn_backends: Dict[str, type[AttentionBackend]],
+        dtype: torch.dtype,
+        threads_per_gpu: Optional[int] = None,
+        max_staging_memory_gb: float = DEFAULT_MAX_STAGING_MEMORY_GB,
+        root_dir: str = "/tmp/shared-kv"
+    ):
+
+        super().__init__(model_name, tp_size, tp_rank, dtype,
```

Comment:
You currently pass in the dtype from the vllm config, which I think can be "auto".
I think you need to take the actual dtype from the `kv_caches` tensors.